<a href="https://colab.research.google.com/github/Dekamonia/MachineLearning/blob/main/lab1/lab01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [5]:
import pandas as pd

In [6]:
df = pd.read_csv("/content/task_data/CarPrice_Assignment.csv")

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [7]:
df.columns

Index(['car_ID', 'symboling', 'CarName', 'fueltype', 'aspiration',
       'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'wheelbase',
       'carlength', 'carwidth', 'carheight', 'curbweight', 'enginetype',
       'cylindernumber', 'enginesize', 'fuelsystem', 'boreratio', 'stroke',
       'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg',
       'price'],
      dtype='object')

In [8]:
df.head()

car_ID  symboling                   CarName fueltype aspiration doornumber  \
0       1          3        alfa-romero giulia      gas        std        two   
1       2          3       alfa-romero stelvio      gas        std        two   
2       3          1  alfa-romero Quadrifoglio      gas        std        two   
3       4          2               audi 100 ls      gas        std       four   
4       5          2                audi 100ls      gas        std       four   

       carbody drivewheel enginelocation  wheelbase  ...  enginesize  \
0  convertible        rwd          front       88.6  ...         130   
1  convertible        rwd          front       88.6  ...         130   
2    hatchback        rwd          front       94.5  ...         152   
3        sedan        fwd          front       99.8  ...         109   
4        sedan        4wd          front       99.4  ...         136   

   fuelsystem  boreratio  stroke compressionratio horsepower  peakrpm citympg  \
0        mpfi       3.47    2.68              9.0        111     5000      21   
1        mpfi       3.47    2.68              9.0        111     5000      21   
2        mpfi       2.68    3.47              9.0        154     5000      19   
3        mpfi       3.19    3.40             10.0        102     5500      24   
4        mpfi       3.19    3.40              8.0        115     5500      18   

   highwaympg    price  
0          27  13495.0  
1          27  16500.0  
2          26  16500.0  
3          30  13950.0  
4          22  17450.0  

[5 rows x 26 columns]

In [9]:
df.tail()

car_ID  symboling          CarName fueltype aspiration doornumber  \
200     201         -1  volvo 145e (sw)      gas        std       four   
201     202         -1      volvo 144ea      gas      turbo       four   
202     203         -1      volvo 244dl      gas        std       four   
203     204         -1        volvo 246   diesel      turbo       four   
204     205         -1      volvo 264gl      gas      turbo       four   

    carbody drivewheel enginelocation  wheelbase  ...  enginesize  fuelsystem  \
200   sedan        rwd          front      109.1  ...         141        mpfi   
201   sedan        rwd          front      109.1  ...         141        mpfi   
202   sedan        rwd          front      109.1  ...         173        mpfi   
203   sedan        rwd          front      109.1  ...         145         idi   
204   sedan        rwd          front      109.1  ...         141        mpfi   

     boreratio  stroke compressionratio horsepower  peakrpm citympg  \
200       3.78    3.15              9.5        114     5400      23   
201       3.78    3.15              8.7        160     5300      19   
202       3.58    2.87              8.8        134     5500      18   
203       3.01    3.40             23.0        106     4800      26   
204       3.78    3.15              9.5        114     5400      19   

     highwaympg    price  
200          28  16845.0  
201          25  19045.0  
202          23  21485.0  
203          27  22470.0  
204          25  22625.0  

[5 rows x 26 columns]

In [10]:
df.describe()

car_ID   symboling   wheelbase   carlength    carwidth   carheight  \
count  205.000000  205.000000  205.000000  205.000000  205.000000  205.000000   
mean   103.000000    0.834146   98.756585  174.049268   65.907805   53.724878   
std     59.322565    1.245307    6.021776   12.337289    2.145204    2.443522   
min      1.000000   -2.000000   86.600000  141.100000   60.300000   47.800000   
25%     52.000000    0.000000   94.500000  166.300000   64.100000   52.000000   
50%    103.000000    1.000000   97.000000  173.200000   65.500000   54.100000   
75%    154.000000    2.000000  102.400000  183.100000   66.900000   55.500000   
max    205.000000    3.000000  120.900000  208.100000   72.300000   59.800000   

        curbweight  enginesize   boreratio      stroke  compressionratio  \
count   205.000000  205.000000  205.000000  205.000000        205.000000   
mean   2555.565854  126.907317    3.329756    3.255415         10.142537   
std     520.680204   41.642693    0.270844    0.313597          3.972040   
min    1488.000000   61.000000    2.540000    2.070000          7.000000   
25%    2145.000000   97.000000    3.150000    3.110000          8.600000   
50%    2414.000000  120.000000    3.310000    3.290000          9.000000   
75%    2935.000000  141.000000    3.580000    3.410000          9.400000   
max    4066.000000  326.000000    3.940000    4.170000         23.000000   

       horsepower      peakrpm     citympg  highwaympg         price  
count  205.000000   205.000000  205.000000  205.000000    205.000000  
mean   104.117073  5125.121951   25.219512   30.751220  13276.710571  
std     39.544167   476.985643    6.542142    6.886443   7988.852332  
min     48.000000  4150.000000   13.000000   16.000000   5118.000000  
25%     70.000000  4800.000000   19.000000   25.000000   7788.000000  
50%     95.000000  5200.000000   24.000000   30.000000  10295.000000  
75%    116.000000  5500.000000   30.000000   34.000000  16503.000000  
max    288.000000  6600.000000   49.000000   54.000000  45400.000000

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета





1.   **Ответ 1 =
                 mean	103.000000	0.834146	98.756585	174.049268	65.907805	53.724878	2555.565854	126.907317	3.329756	3.255415	10.142537	104.117073	5125.121951	25.219512	30.751220	13276.710571.**
2.   **Ответ 2 =
                 min	1.000000	-2.000000	86.600000	141.100000	60.300000	47.800000	1488.000000	61.000000	2.540000	2.070000	7.000000	48.000000	4150.000000	13.000000	16.000000	5118.000000
                 max	205.000000	3.000000	120.900000	208.100000	72.300000	59.800000	4066.000000	326.000000	3.940000	4.170000	23.000000	288.000000	6600.000000	49.000000	54.000000	45400.000000**



##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [11]:
df.car_ID = df.car_ID.astype("int")
df.symboling = df.symboling.astype("int")
df.CarName = df.CarName.astype("category")
df.fueltype = df.fueltype.astype("category")
df.aspiration = df.aspiration.astype("category")
df.doornumber = df.doornumber.astype("category")
df.carbody = df.carbody.astype("category")
df.drivewheel = df.drivewheel.astype("category")
df.enginelocation = df.enginelocation.astype("category")
df.wheelbase = df.wheelbase.astype("float")
df.carlength = df.carlength.astype("float")
df.carwidth = df.carwidth.astype("float")
df.carheight = df.carheight.astype("float")
df.curbweight = df.curbweight.astype("int")
df.enginetype = df.enginetype.astype("category")
df.cylindernumber = df.cylindernumber.astype("category")
df.enginesize = df.enginesize.astype("int")
df.fuelsystem = df.fuelsystem.astype("category")
df.boreratio = df.boreratio.astype("float")
df.stroke = df.stroke.astype("float")
df.compressionratio = df.compressionratio.astype("float")
df.horsepower = df.horsepower.astype("int")
df.peakrpm = df.peakrpm.astype("int")
df.citympg = df.citympg.astype("int")
df.highwaympg = df.highwaympg.astype("int")
df.price = df.price.astype("int")
#Вывести типы колонок тут
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   car_ID            205 non-null    int64   
 1   symboling         205 non-null    int64   
 2   CarName           205 non-null    category
 3   fueltype          205 non-null    category
 4   aspiration        205 non-null    category
 5   doornumber        205 non-null    category
 6   carbody           205 non-null    category
 7   drivewheel        205 non-null    category
 8   enginelocation    205 non-null    category
 9   wheelbase         205 non-null    float64 
 10  carlength         205 non-null    float64 
 11  carwidth          205 non-null    float64 
 12  carheight         205 non-null    float64 
 13  curbweight        205 non-null    int64   
 14  enginetype        205 non-null    category
 15  cylindernumber    205 non-null    category
 16  enginesize        205 non-

Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [12]:
df[df["fueltype"] == "gas"]

car_ID  symboling                   CarName fueltype aspiration  \
0         1          3        alfa-romero giulia      gas        std   
1         2          3       alfa-romero stelvio      gas        std   
2         3          1  alfa-romero Quadrifoglio      gas        std   
3         4          2               audi 100 ls      gas        std   
4         5          2                audi 100ls      gas        std   
..      ...        ...                       ...      ...        ...   
199     200         -1              volvo diesel      gas      turbo   
200     201         -1           volvo 145e (sw)      gas        std   
201     202         -1               volvo 144ea      gas      turbo   
202     203         -1               volvo 244dl      gas        std   
204     205         -1               volvo 264gl      gas      turbo   

    doornumber      carbody drivewheel enginelocation  wheelbase  ...  \
0          two  convertible        rwd          front       88.6  ...   
1          two  convertible        rwd          front       88.6  ...   
2          two    hatchback        rwd          front       94.5  ...   
3         four        sedan        fwd          front       99.8  ...   
4         four        sedan        4wd          front       99.4  ...   
..         ...          ...        ...            ...        ...  ...   
199       four        wagon        rwd          front      104.3  ...   
200       four        sedan        rwd          front      109.1  ...   
201       four        sedan        rwd          front      109.1  ...   
202       four        sedan        rwd          front      109.1  ...   
204       four        sedan        rwd          front      109.1  ...   

     enginesize  fuelsystem  boreratio  stroke compressionratio horsepower  \
0           130        mpfi       3.47    2.68              9.0        111   
1           130        mpfi       3.47    2.68              9.0        111   
2           152        mpfi       2.68    3.47              9.0        154   
3           109        mpfi       3.19    3.40             10.0        102   
4           136        mpfi       3.19    3.40              8.0        115   
..          ...         ...        ...     ...              ...        ...   
199         130        mpfi       3.62    3.15              7.5        162   
200         141        mpfi       3.78    3.15              9.5        114   
201         141        mpfi       3.78    3.15              8.7        160   
202         173        mpfi       3.58    2.87              8.8        134   
204         141        mpfi       3.78    3.15              9.5        114   

     peakrpm citympg  highwaympg  price  
0       5000      21          27  13495  
1       5000      21          27  16500  
2       5000      19          26  16500  
3       5500      24          30  13950  
4       5500      18          22  17450  
..       ...     ...         ...    ...  
199     5100      17          22  18950  
200     5400      23          28  16845  
201     5300      19          25  19045  
202     5500      18          23  21485  
204     5400      19          25  22625  

[185 rows x 26 columns]

In [13]:
df[df["car_ID"] < 31]

car_ID  symboling                    CarName fueltype aspiration  \
0        1          3         alfa-romero giulia      gas        std   
1        2          3        alfa-romero stelvio      gas        std   
2        3          1   alfa-romero Quadrifoglio      gas        std   
3        4          2                audi 100 ls      gas        std   
4        5          2                 audi 100ls      gas        std   
5        6          2                   audi fox      gas        std   
6        7          1                 audi 100ls      gas        std   
7        8          1                  audi 5000      gas        std   
8        9          1                  audi 4000      gas      turbo   
9       10          0        audi 5000s (diesel)      gas      turbo   
10      11          2                   bmw 320i      gas        std   
11      12          0                   bmw 320i      gas        std   
12      13          0                     bmw x1      gas        std   
13      14          0                     bmw x3      gas        std   
14      15          1                     bmw z4      gas        std   
15      16          0                     bmw x4      gas        std   
16      17          0                     bmw x5      gas        std   
17      18          0                     bmw x3      gas        std   
18      19          2           chevrolet impala      gas        std   
19      20          1      chevrolet monte carlo      gas        std   
20      21          0        chevrolet vega 2300      gas        std   
21      22          1              dodge rampage      gas        std   
22      23          1        dodge challenger se      gas        std   
23      24          1                 dodge d200      gas      turbo   
24      25          1          dodge monaco (sw)      gas        std   
25      26          1         dodge colt hardtop      gas        std   
26      27          1            dodge colt (sw)      gas        std   
27      28          1       dodge coronet custom      gas      turbo   
28      29         -1          dodge dart custom      gas        std   
29      30          3  dodge coronet custom (sw)      gas      turbo   

   doornumber      carbody drivewheel enginelocation  wheelbase  ...  \
0         two  convertible        rwd          front       88.6  ...   
1         two  convertible        rwd          front       88.6  ...   
2         two    hatchback        rwd          front       94.5  ...   
3        four        sedan        fwd          front       99.8  ...   
4        four        sedan        4wd          front       99.4  ...   
5         two        sedan        fwd          front       99.8  ...   
6        four        sedan        fwd          front      105.8  ...   
7        four        wagon        fwd          front      105.8  ...   
8        four        sedan        fwd          front      105.8  ...   
9         two    hatchback        4wd          front       99.5  ...   
10        two        sedan        rwd          front      101.2  ...   
11       four        sedan        rwd          front      101.2  ...   
12        two        sedan        rwd          front      101.2  ...   
13       four        sedan        rwd          front      101.2  ...   
14       four        sedan        rwd          front      103.5  ...   
15       four        sedan        rwd          front      103.5  ...   
16        two        sedan        rwd          front      103.5  ...   
17       four        sedan        rwd          front      110.0  ...   
18        two    hatchback        fwd          front       88.4  ...   
19        two    hatchback        fwd          front       94.5  ...   
20       four        sedan        fwd          front       94.5  ...   
21        two    hatchback        fwd          front       93.7  ...   
22        two    hatchback        fwd          front       93.7  ...   
23        two    hatchback        fwd      

In [14]:
df[df["carbody"] == "hatchback"]

car_ID  symboling                   CarName fueltype aspiration  \
2         3          1  alfa-romero Quadrifoglio      gas        std   
9        10          0       audi 5000s (diesel)      gas      turbo   
18       19          2          chevrolet impala      gas        std   
19       20          1     chevrolet monte carlo      gas        std   
21       22          1             dodge rampage      gas        std   
..      ...        ...                       ...      ...        ...   
175     176         -1             toyota corona      gas        std   
177     178         -1            toyota mark ii      gas        std   
178     179          3   toyota corolla liftback      gas        std   
179     180          3             toyota corona      gas        std   
190     191          3                 vw rabbit      gas        std   

    doornumber    carbody drivewheel enginelocation  wheelbase  ...  \
2          two  hatchback        rwd          front       94.5  ...   
9          two  hatchback        4wd          front       99.5  ...   
18         two  hatchback        fwd          front       88.4  ...   
19         two  hatchback        fwd          front       94.5  ...   
21         two  hatchback        fwd          front       93.7  ...   
..         ...        ...        ...            ...        ...  ...   
175       four  hatchback        fwd          front      102.4  ...   
177       four  hatchback        fwd          front      102.4  ...   
178        two  hatchback        rwd          front      102.9  ...   
179        two  hatchback        rwd          front      102.9  ...   
190        two  hatchback        fwd          front       94.5  ...   

     enginesize  fuelsystem  boreratio  stroke compressionratio horsepower  \
2           152        mpfi       2.68    3.47             9.00        154   
9           131        mpfi       3.13    3.40             7.00        160   
18           61        2bbl       2.91    3.03             9.50         48   
19           90        2bbl       3.03    3.11             9.60         70   
21           90        2bbl       2.97    3.23             9.41         68   
..          ...         ...        ...     ...              ...        ...   
175         122        mpfi       3.31    3.54             8.70         92   
177         122        mpfi       3.31    3.54             8.70         92   
178         171        mpfi       3.27    3.35             9.30        161   
179         171        mpfi       3.27    3.35             9.30        161   
190         109        mpfi       3.19    3.40             8.50         90   

     peakrpm citympg  highwaympg  price  
2       5000      19          26  16500  
9       5500      16          22  17859  
18      5100      47          53   5151  
19      5400      38          43   6295  
21      5500      37          41   5572  
..       ...     ...         ...    ...  
175     4200      27          32   9988  
177     4200      27          32  11248  
178     5200      20          24  16558  
179     5200      19          24  15998  
190     5500      24          29   9980  

[70 rows x 26 columns]

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [17]:
df.groupby('carbody').mean()

<ipython-input-17-6cf2a7fb3213>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby('carbody').mean()


car_ID  symboling   wheelbase   carlength   carwidth  \
carbody                                                                 
convertible   94.666667   2.833333   92.700000  170.383333  65.583333   
hardtop      125.875000   1.875000   98.500000  176.937500  66.612500   
hatchback     93.842857   1.585714   95.535714  166.592857  65.247143   
sedan        102.072917   0.333333  100.619792  177.585417  66.260417   
wagon        126.880000  -0.160000  102.156000  181.304000  66.256000   

             carheight   curbweight  enginesize  boreratio    stroke  \
carbody                                                                
convertible  51.433333  2801.666667  157.166667   3.491667  3.043333   
hardtop      52.850000  2810.625000  176.250000   3.608750  3.322500   
hatchback    52.108571  2348.185714  114.400000   3.244571  3.278143   
sedan        54.337500  2610.593750  130.822917   3.338542  3.267292   
wagon        56.728000  2784.240000  123.840000   3.406400  3.175600   

             compressionratio  horsepower      peakrpm    citympg  highwaympg  \
carbody                                                                         
convertible          8.933333  131.666667  5158.333333  20.500000   26.000000   
hardtop             10.725000  142.250000  5031.250000  21.625000   27.250000   
hatchback            9.027429  101.171429  5230.714286  26.314286   32.171429   
sedan               10.937500  103.104167  5081.770833  25.322917   30.833333   
wagon               10.316000   97.440000  5018.000000  24.040000   28.720000   

                    price  
carbody                    
convertible  21890.500000  
hardtop      22208.500000  
hatchback    10376.642857  
sedan        14344.260417  
wagon        12371.960000

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.

In [16]:
#Агрегация